### PR TITLE
server: fix the race on (*Server).clusterID

### LINKED ISSUE #7772: Data race on (*server).clusterID
```shell
2024-01-29T08:26:32.7122217Z WARNING: DATA RACE
2024-01-29T08:26:32.7122514Z Write at 0x00c002df00a8 by goroutine 2830:
2024-01-29T08:26:32.7122956Z   github.com/tikv/pd/server.(*Server).startServer()
2024-01-29T08:26:32.7123500Z       /home/runner/work/pd/pd/server/server.go:429 +0x104
2024-01-29T08:26:32.7123943Z   github.com/tikv/pd/server.(*Server).Run()
2024-01-29T08:26:32.7124519Z       /home/runner/work/pd/pd/server/server.go:601 +0x1e6
2024-01-29T08:26:32.7124977Z   github.com/tikv/pd/tests.(*TestServer).Run()
2024-01-29T08:26:32.7125545Z       /home/runner/work/pd/pd/tests/cluster.go:125 +0x1fd
2024-01-29T08:26:32.7126062Z   github.com/tikv/pd/tests.(*TestCluster).RunServer.func1()
2024-01-29T08:26:32.7126631Z       /home/runner/work/pd/pd/tests/cluster.go:575 +0x44
2024-01-29T08:26:32.7126956Z 
2024-01-29T08:26:32.7127125Z Previous read at 0x00c002df00a8 by goroutine 2955:
2024-01-29T08:26:32.7127615Z   github.com/tikv/pd/server.(*GrpcServer).errorHeader()
2024-01-29T08:26:32.7128207Z       /home/runner/work/pd/pd/server/grpc_service.go:2301 +0x311
2024-01-29T08:26:32.7128779Z   github.com/tikv/pd/server.(*GrpcServer).wrapErrorToHeader()
2024-01-29T08:26:32.7129400Z       /home/runner/work/pd/pd/server/grpc_service.go:2293 +0x199
2024-01-29T08:26:32.7129939Z   github.com/tikv/pd/server.(*GrpcServer).GetClusterInfo()
2024-01-29T08:26:32.7130540Z       /home/runner/work/pd/pd/server/grpc_service.go:267 +0x11a
2024-01-29T08:26:32.7131176Z   github.com/pingcap/kvproto/pkg/pdpb._PD_GetClusterInfo_Handler.func1()
2024-01-29T08:26:32.7132302Z       /home/runner/go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20231226064240-4f28b82c7860/pkg/pdpb/pdpb.pb.go:9572 +0x82
2024-01-29T08:26:32.7133216Z   go.etcd.io/etcd/etcdserver/api/v3rpc.Server.ChainUnaryServer.func5.1()
2024-01-29T08:26:32.7134402Z       /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:31 +0x223
2024-01-29T08:26:32.7135539Z   github.com/grpc-ecosystem/go-grpc-prometheus.init.(*ServerMetrics).UnaryServerInterceptor.func3()
2024-01-29T08:26:32.7136700Z       /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0xbb
2024-01-29T08:26:32.7137560Z   go.etcd.io/etcd/etcdserver/api/v3rpc.Server.ChainUnaryServer.func5.1()
2024-01-29T08:26:32.7138724Z       /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x182
2024-01-29T08:26:32.7139678Z   go.etcd.io/etcd/etcdserver/api/v3rpc.Server.newUnaryInterceptor.func4()
2024-01-29T08:26:32.7140885Z       /home/runner/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/etcdserver/api/v3rpc/interceptor.go:70 +0x2c1
2024-01-29T08:26:32.7141853Z   go.etcd.io/etcd/etcdserver/api/v3rpc.Server.ChainUnaryServer.func5.1()
2024-01-29T08:26:32.7143006Z       /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x182
2024-01-29T08:26:32.7143980Z   go.etcd.io/etcd/etcdserver/api/v3rpc.Server.newLogUnaryInterceptor.func3()
2024-01-29T08:26:32.7145187Z       /home/runner/go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20220915004622-85b640cee793/etcdserver/api/v3rpc/interceptor.go:77 +0xc8
2024-01-29T08:26:32.7146419Z   go.etcd.io/etcd/etcdserver/api/v3rpc.Server.ChainUnaryServer.func5()
2024-01-29T08:26:32.7147594Z       /home/runner/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:39 +0x32e
2024-01-29T08:26:32.7148512Z   github.com/pingcap/kvproto/pkg/pdpb._PD_GetClusterInfo_Handler()
2024-01-29T08:26:32.7149603Z       /home/runner/go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20231226064240-4f28b82c7860/pkg/pdpb/pdpb.pb.go:9574 +0x1c3
2024-01-29T08:26:32.7150404Z   google.golang.org/grpc.(*Server).processUnaryRPC()
2024-01-29T08:26:32.7151198Z       /home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1024 +0x1502
2024-01-29T08:26:32.7151805Z   google.golang.org/grpc.(*Server).handleStream()
2024-01-29T08:26:32.7152493Z       /home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:1313 +0xfd2
2024-01-29T08:26:32.7153123Z   google.golang.org/grpc.(*Server).serveStreams.func1.1()
2024-01-29T08:26:32.7153820Z       /home/runner/go/pkg/mod/google.golang.org/grpc@v1.26.0/server.go:722 +0xfa
```

https://github.com/tikv/pd/actions/runs/7692536672/job/20961055287?pr=7737

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: Close #7772.

### What is changed and how does it work?

```commit-message
Fix the race on `(*Server).clusterID`.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by rleungx — APPROVED** [review]

File: server/grpc_service.go [hunk 1]
```diff
@@ -2276,17 +2276,19 @@ func (s *GrpcServer) validateRoleInRequest(ctx context.Context, header *pdpb.Req
 		}
 		*allowFollower = true
 	}
-	if header.GetClusterId() != s.clusterID {
-		return status.Errorf(codes.FailedPrecondition, "mismatch cluster id, need %d but got %d", s.clusterID, header.GetClusterId())
+	clusterID := s.ClusterID()
+	if header.GetClusterId() != clusterID {
+		return status.Errorf(codes.FailedPrecondition, "mismatch cluster id, need %d but got %d", clusterID, header.GetClusterId())
```
Comment (HuSharp) [inline server/grpc_service.go:2281]: a Nitpick :) can we keep consistent with above code?
```suggestion
	if clusterID := s.ClusterID(); header.GetClusterId() != clusterID {
		return status.Errorf(codes.FailedPrecondition, "mismatch cluster id, need %d but got %d", clusterID, header.GetClusterId())
```

**Review by HuSharp — COMMENTED** [review]
I have tested the data race in my own repository(**which run 13 times in `/mcs/tso` pls notice makefile** ) and there may still be some problems https://github.com/HuSharp/pd/actions/runs/7693470352/job/20962291846?pr=5

**Comment by JmPotato** [comment]
> * I have tested the data race in my own repository(**which runs 13 times in `/mcs/tso` pls notice makefile** ) and with repairing `clusterID` there may still be some problems [HuSharp/pd/actions/runs/7693470352/job/20962291846?pr=5](https://github.com/HuSharp/pd/actions/runs/7693470352/job/20962291846?pr=5)
> * do we need to cp this pr?

The data race appears to be a separate issue. I would prefer to address it at a later time, if we encounter it again.

**Review by HuSharp — APPROVED** [review]

**Pull request merged**